### PR TITLE
Updating Event Argument

### LINKED ISSUE #26789: `sales_order_state_change_before` Event should give access to order object
### Summary
When placing an observer on `sales_order_state_change_before`, it's not possible to get any data from the actual order object.

### Examples
see `\Magento\Sales\Model\Service\OrderService->setState()`:

Current code is:
`$this->eventManager->dispatch(
            'sales_order_state_change_before',
            ['order' => $this, 'transport' => $transport]
        );`

Because `$this` is passed to the observer, and not `$order`, there's no way to get order data in the observer.

### Proposed solution
Change event dispatch either to `'order' => $order`, or add `'order_object'` to have it backwards compatible.

`$this->eventManager->dispatch(
            'sales_order_state_change_before',
            ['order' => $this, 'transport' => $transport, 'order_object' => $order]
        );`

### PR DESCRIPTION
**Pull request by manavluhar**
### Description (*)
Updating event argument for sales_order_state_change_before event

### Fixed Issues (if relevant)
1. Fixes magento/magento2#26789

### Manual testing scenarios (*)
1. place an observer on sales_order_state_change_before and try to get the data from the actual object

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [x] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by manavluhar** [comment]
@magento run all tests

**Comment by manavluhar** [comment]
@magento run Static Tests, Functional Tests B2B

**Comment by manavluhar** [comment]
@magento run all tests

File: app/code/Magento/Sales/Model/Service/OrderService.php [hunk 1]
```diff
@@ -267,9 +267,12 @@ public function setState(
             ]
         );
 
+        /**
+        * Event argument `transport` is @deprecated. Use `transportObject` instead.
+        */
         $this->eventManager->dispatch(
             'sales_order_state_change_before',
-            ['order' => $this, 'transport' => $transport]
+            ['order' => $this, 'transport' => $transport, 'transportObject' => $transport]
```
Comment (fooman) [inline app/code/Magento/Sales/Model/Service/OrderService.php]:
```suggestion
            ['order' => $this, 'transport' => $transport, 'order_object' => $order]
```
To address the concern in the reported issue the order object should be made available via the event.

**Comment by manavluhar** [comment]
@magento run all tests

**Comment by manavluhar** [comment]
@magento run all tests

**Comment by manavluhar** [comment]
@magento run all tests

**Comment by manavluhar** [comment]
@magento run Functional Tests EE, Functional Tests CE, Functional Tests B2B

**Comment by manavluhar** [comment]
@magento run Functional Tests EE

**Comment by manavluhar** [comment]
@magento run Functional Tests EE, Functional Tests CE, Functional Tests B2B

**Comment by manavluhar** [comment]
@magento run Functional Tests CE, Functional Tests EE, Functional Tests B2B

**Comment by manavluhar** [comment]
@magento run Functional Tests EE

**Comment by manavluhar** [comment]
@magento run all tests

**Comment by manavluhar** [comment]
@magento run Functional Tests CE, Functional Tests EE, Functional Tests B2B

**Comment by manavluhar** [comment]
@magento run Functional Tests CE

**Comment by manavluhar** [comment]
@magento run Functional Tests CE, Functional Tests EE, Functional Tests B2B

**Comment by manavluhar** [comment]
@magento run Functional Tests CE, Functional Tests EE, Functional Tests B2B

**Comment by manavluhar** [comment]
@magento run all tests

**Comment by manavluhar** [comment]
@magento run all tests

**Comment by manavluhar** [comment]
@magento run all tests

**Comment by manavluhar** [comment]
@fooman Can you please review this Pull Request again?

**Comment by manavluhar** [comment]
@magento run all tests

**Comment by manavluhar** [comment]
@fooman🙌 Can you please review this  PR?

**Comment by fooman** [comment]
@manavluhar Code looks okay to me, but I am currently not part of the maintainer team which will need to also take a look. One thing which is not accurate is the comment

```
        /**
        * Event argument `transport` is @deprecated. Use `transportObject` instead.
        */
```

I do not see anything being deprecated so I would remove the comment completely.

File: app/code/Magento/Sales/Model/Service/OrderService.php [hunk 2]
```diff
@@ -267,9 +267,12 @@ public function setState(
             ]
         );
 
+        /**
+        * Event argument `transport` is @deprecated. Use `transportObject` instead.
+        */
```
Comment (fooman) [inline app/code/Magento/Sales/Model/Service/OrderService.php]:
```suggestion
```

**Review by ghost — CHANGES_REQUESTED** [review]

**Comment by manavluhar** [comment]
> @manavluhar Code looks okay to me, but I am currently not part of the maintainer team which will need to also take a look. One thing which is not accurate is the comment
> 
> ```
>         /**
>         * Event argument `transport` is @deprecated. Use `transportObject` instead.
>         */
> ```
> 
> I do not see anything being deprecated so I would remove the comment completely.

Thank you for the review Sir, can you please approve your review so this pull request can proceed further?

**Comment by manavluhar** [comment]
@magento run all tests

**Comment by manavluhar** [comment]
@magento run all tests

**Comment by manavluhar** [comment]
@fooman can you please review this pr?

**Comment by fooman** [comment]
@manavluhar looks good to me but you would need someone from the maintainer group give the go-ahead to move this PR along.

**Review by ghost — APPROVED** [review]

**Comment by manavluhar** [comment]
@magento run all tests

**Comment by engcom-Lima** [comment]
@magento run Functional Tests B2B, Integration Tests

**Comment by engcom-Lima** [comment]
@magento run Integration Tests

**Comment by engcom-Lima** [comment]
@magento give me 2.4.3-p1 with B2b instance

**Comment by engcom-Charlie** [comment]
Hi @manavluhar,

Thank you for your contribution!

**In order to reproduce the issue and test the solution, I have implemented below code base.:**

1. Added sales_order_state_change_before“ event-observer entry in app/code/Magento/Sales/etc/events.xml
2. Created custom observer for above event
3. Created new custom controller and load Order and OrderService object in it. Load the order and called $orderService->setState function.
4. Added debugger points in controller and in observer, in order to reproduce the issue.
5. Found that we are not getting Order object’s data. Hence issue is reproducible.
image

:heavy_check_mark: QA Passed
6. Added the PR changes and observed that after PR solution we are getting Order object data correctly.
image

Also tested all states of order from order grid in admin. 

Since the issue is fix and the build is green moving it for merging.

Thank you!

**Pull request merged**